mypy_django_plugin/transformers/managers.py

```diff
@@ -183,112 +247,182 @@ def create_new_manager_class_from_from_queryset_method(ctx: DynamicClassDefConte
-    # Stash the queryset fullname which was passed to .from_queryset
-    # So that our 'resolve_manager_method' attribute hook can fetch the method from that QuerySet class
-    new_manager_info.metadata["django"] = {"from_queryset_manager": derived_queryset_fullname}
+    queryset_info = queryset_sym.node
+    assert isinstance(queryset_info, TypeInfo)
 
     if len(ctx.call.args) > 1:
         expr = ctx.call.args[1]
         assert isinstance(expr, StrExpr)
-        custom_manager_generated_name = expr.value
+        manager_class_name = expr.value
     else:
-        custom_manager_generated_name = base_manager_info.name + "From" + derived_queryset_info.name
+        manager_class_name = manager_base.name + "From" + queryset_info.name
 
-    custom_manager_generated_fullname = ".".join(["django.db.models.manager", custom_manager_generated_name])
-    base_manager_info.metadata.setdefault("from_queryset_managers", {})
-    base_manager_info.metadata["from_queryset_managers"][custom_manager_generated_fullname] = new_manager_info.fullname
+    manager_base_instance = fill_typevars(manager_base)
+    assert isinstance(manager_base_instance, Instance)
+    # Create a new `TypeInfo` instance for the manager type
+    new_manager_info = semanal_api.basic_new_typeinfo(
+        name=manager_class_name, basetype_or_fallback=manager_base_instance, line=ctx.call.line
+    )
+    new_manager_info.type_vars = manager_base.type_vars
+    new_manager_info.line = ctx.call.line
+    new_manager_info.defn.type_vars = manager_base.defn.type_vars
+    new_manager_info.defn.line = ctx.call.line
+    new_manager_info.metaclass_type = new_manager_info.calculate_metaclass_type()
 
-    # So that the plugin will reparameterize the manager when it is constructed inside of a Model definition
-    helpers.add_new_manager_base(semanal_api, new_manager_info.fullname)
+    merge_queryset_into_manager(
+        api=semanal_api,
+        manager_info=new_manager_info,
+        queryset_info=queryset_info,
+        manager_base=manager_base,
+        class_name=manager_class_name,
+    )
 
-    class_def_context = ClassDefContext(cls=new_manager_info.defn, reason=ctx.call, api=semanal_api)
-    self_type = fill_typevars(new_manager_info)
-    assert isinstance(self_type, Instance)
+    symbol_kind = semanal_api.current_symbol_kind()
+    # Annotate the module variable as `<Variable>: Type[<NewManager[Any]>]` as the model
+    # type won't be defined on variable level.
+    var = Var(
+        name=ctx.name,
+        type=TypeType(Instance(new_manager_info, [AnyType(TypeOfAny.from_omitted_generics)])),
+    )
+    var.info = new_manager_info
+    var._fullname = f"{semanal_api.cur_mod_id}.{ctx.name}"
+    var.is_inferred = True
+    # Note: Order of `add_symbol_table_node` calls matter. Case being if
+    # `ctx.name == new_manager_info.name`, then we'd _only_ like the type and not the
+    # `Var` to exist..
```

Comment:
Took me a while to understand that in the case of `ctx.name == new_manager_info.name` the second call will basically override the changes made by the first call. So maybe explicitly state that?

Reply:
Sure, we can do that

Reply:
Let me know if it's more obvious with the updated comment